File: packages/contracts/src/observation.ts
import { z } from "zod";
import { stopSchema } from "./stops.js";
import { OBSERVATION_SCHEMA_VERSION } from "./versions.js";

const idSchema = z.string().min(1);
const timestampSchema = z.int().nonnegative();

/** A vehicle is either stationary at a node or committed to its current edge. */
export const vehiclePositionSchema = z.discriminatedUnion("kind", [
  z.object({ kind: z.literal("atNode"), nodeId: idSchema }),
  z.object({
    kind: z.literal("onEdge"),
    edgeId: idSchema,
    fromNodeId: idSchema,
    toNodeId: idSchema,
    arrivalTimeMs: timestampSchema,
  }),
]);

/** A planned stop with its host-computed arrival time along the route. */
export const candidateStopSchema = stopSchema.extend({ plannedArrivalTimeMs: timestampSchema });

export const observedVehicleSchema = z.object({
  id: idSchema,
  capacity: z.int().positive(),
  position: vehiclePositionSchema,
  onboardRequestIds: z.array(idSchema),
  /** Remaining committed stops in service order, with planned arrival times. */
  stops: z.array(candidateStopSchema),
});

export const REQUEST_PHASES = ["waiting", "assigned", "onboard", "completed"] as const;
export const requestPhaseSchema = z.enum(REQUEST_PHASES);

/** A released request. Unreleased requests never appear in an observation. */
export const observedRequestSchema = z.object({
  id: idSchema,
  requestTimeMs: timestampSchema,
  originNodeId: idSchema,
  destinationNodeId: idSchema,
  phase: requestPhaseSchema,
  /** Shortest direct travel time on the fixed graph, ms. */
  directTravelTimeMs: z.int().nonnegative(),
});

/**
 * A legal insertion candidate. It carries the complete remaining stop list of
 * one vehicle. Candidates never carry pain, cost, or a cost-based rank, and
 * their order is fixed by (vehicleId, candidate id) rather than quality.
 */
export const candidateSchema = z.object({
  id: idSchema,
  vehicleId: idSchema,
  stops: z.array(candidateStopSchema).min(1),
});

export const observationSchema = z.object({
  schemaVersion: z.literal(OBSERVATION_SCHEMA_VERSION),
  scenarioId: idSchema,
  /** Monotonic state version; actions must name the version they were decided on. */
  stateVersion: z.int().nonnegative(),
  nowMs: timestampSchema,
  /** The request whose pickup and drop-off must be inserted by this decision. */
  decisionRequestId: idSchema,
  vehicles: z.array(observedVehicleSchema).min(1),
  requests: z.array(observedRequestSchema),
  candidates: z.array(candidateSchema),
});

export type VehiclePosition = z.infer<typeof vehiclePositionSchema>;
export type ObservedVehicle = z.infer<typeof observedVehicleSchema>;
export type RequestPhase = z.infer<typeof requestPhaseSchema>;
export type ObservedRequest = z.infer<typeof observedRequestSchema>;
export type CandidateStop = z.infer<typeof candidateStopSchema>;
export type Candidate = z.infer<typeof candidateSchema>;
export type Observation = z.infer<typeof observationSchema>;

File: swift/Sources/Bus20Core/Observation.swift
import Foundation

/// Mirrors `bus20-observation/1` from `@bus20/contracts`. Only fields the
/// reference needs are modelled; unknown fields are ignored on decode.
public struct Observation: Codable, Equatable {
    public let schemaVersion: String
    public let scenarioId: String
    public let stateVersion: Int
    public let nowMs: Int
    public let decisionRequestId: String
    public let vehicles: [ObservedVehicle]
    public let requests: [ObservedRequest]
    public let candidates: [Candidate]

    public init(schemaVersion: String, scenarioId: String, stateVersion: Int, nowMs: Int,
                decisionRequestId: String, vehicles: [ObservedVehicle],
                requests: [ObservedRequest], candidates: [Candidate]) {
        self.schemaVersion = schemaVersion
        self.scenarioId = scenarioId
        self.stateVersion = stateVersion
        self.nowMs = nowMs
        self.decisionRequestId = decisionRequestId
        self.vehicles = vehicles
        self.requests = requests
        self.candidates = candidates
    }
}

public enum StopKind: String, Codable, Equatable {
    case pickup
    case dropoff
}

/// A planned stop with the host-computed arrival time along the route.
public struct TimedStop: Codable, Equatable {
    public let requestId: String
    public let kind: StopKind
    public let nodeId: String
    public let plannedArrivalTimeMs: Int

    public init(requestId: String, kind: StopKind, nodeId: String, plannedArrivalTimeMs: Int) {
        self.requestId = requestId
        self.kind = kind
        self.nodeId = nodeId
        self.plannedArrivalTimeMs = plannedArrivalTimeMs
    }
}

public struct ObservedVehicle: Codable, Equatable {
    public let id: String
    public let capacity: Int
    public let onboardRequestIds: [String]
    /// Remaining committed stops with planned arrival times.
    public let stops: [TimedStop]

    public init(id: String, capacity: Int, onboardRequestIds: [String], stops: [TimedStop]) {
        self.id = id
        self.capacity = capacity
        self.onboardRequestIds = onboardRequestIds
        self.stops = stops
    }
}

public struct ObservedRequest: Codable, Equatable {
    public let id: String
    public let requestTimeMs: Int
    public let originNodeId: String
    public let destinationNodeId: String
    public let phase: String
    public let directTravelTimeMs: Int

    public init(id: String, requestTimeMs: Int, originNodeId: String, destinationNodeId: String,
                phase: String, directTravelTimeMs: Int) {
        self.id = id
        self.requestTimeMs = requestTimeMs
        self.originNodeId = originNodeId
        self.destinationNodeId = destinationNodeId
        self.phase = phase
        self.directTravelTimeMs = directTravelTimeMs
    }
}

public struct Candidate: Codable, Equatable {
    public let id: String
    public let vehicleId: String
    public let stops: [TimedStop]

    public init(id: String, vehicleId: String, stops: [TimedStop]) {
        self.id = id
        self.vehicleId = vehicleId
        self.stops = stops
    }
}

/// Mirrors the `chooseCandidate` form of `bus20-action/1`.
public struct ChooseCandidateAction: Codable, Equatable {
    public let kind = "chooseCandidate"
    public let schemaVersion = "bus20-action/1"
    public let stateVersion: Int
    public let candidateId: String

    enum CodingKeys: String, CodingKey {
        case kind, schemaVersion, stateVersion, candidateId
    }

    public init(stateVersion: Int, candidateId: String) {
        self.stateVersion = stateVersion
        self.candidateId = candidateId
    }
}
